File: Data Structures & Algorithms/maximum-product-subarray/submission-1.java
class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        int[] dpMax = new int[n];
        int[] dpMin = new int[n];
        dpMax[0] = nums[0];
        dpMin[0] = nums[0];
        int res = nums[0];
        for (int i = 1; i < n; i++) {
            int curMax = dpMax[i - 1] * nums[i];
            int curMin = dpMin[i - 1] * nums[i];

            dpMax[i] = Math.max(curMax, Math.max(curMin, nums[i]));
            dpMin[i] = Math.min(curMax, Math.min(curMin, nums[i]));

            res = Math.max(res, dpMax[i]);
        }

        return res; 

    }
}

File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        res = 0
        lo = 0
        hi = len(heights) - 1

        while lo < hi:
            res = max(res, min(heights[lo], heights[hi]) * (hi - lo))
            if heights[lo] < heights[hi]:
                lo += 1
            else:
                hi -= 1
        
        return res